### PR TITLE
Fix `--buildArg` with compose projects; Convert `buildProject` to Typescript

### PR DESCRIPTION
**Pull request by pdcastro**
Resolves: #1053 
Change-type: patch

### CODE REVIEW

File: lib/utils/compose.js [hunk 1]
```diff
@@ -641,8 +365,7 @@ export const pushAndUpdateServiceImages = function (
 		Promise.all(
 			images.map(({ serviceImage, localImage, props, logs }, index) =>
 				Promise.all([
-					// @ts-ignore
-					localImage.inspect().get('Size'),
+					localImage.inspect().then((img) => img.Size),
```
Comment (pdcastro) [inline lib/utils/compose.js:368]: Typing info for `docker-toolbelt` declares native promises, but apparently it actually uses Bluebird promises and the old CLI code was relying on Bluebird being used -- the `.get('Size')` call was Bluebird's `.get`:
http://bluebirdjs.com/docs/api/get.html
This type _vs._ usage incompatibility was ignored with a `"// @ts-ignore"` comment. I've changed the code here and elsewhere not to assume Bluebird, and removed the `ts-ignore` comment.

File: lib/utils/compose.js [hunk 2]
```diff
@@ -85,23 +85,6 @@ export function createProject(composePath, composeStr, projectName = null) {
 	};
 }
 
-/**
```
Comment (pdcastro) [inline lib/utils/compose.js:88]: The deleted code in `compose.js` was converted to Typescript and moved to the pre-existing `compose_ts.ts` file.

File: lib/utils/compose_ts.ts [hunk 3]
```diff
@@ -131,6 +130,372 @@ async function resolveProject(
...
+	arch: string;
+	deviceType: string;
+	emulated: boolean;
+	buildOpts: import('./docker').BuildOpts;
+	inlineLogs?: boolean;
+	convertEol: boolean;
+	dockerfilePath?: string;
+	nogitignore: boolean;
+	multiDockerignore: boolean;
+}): Promise<BuiltImage[]> {
```
Comment (pdcastro) [inline lib/utils/compose_ts.ts:158]: `buildProject` was very large single function written in a `.then` promise chain. Such large functions are harder to test (if we wanted to...) and harder to make sense of, in my opinion. I've split it in some 7 thematic smaller functions, mostly mapping to `.then()` blocks of the old promise chain. Even then, what was left in `buildProject` itself is still of a significant size!

File: lib/utils/compose_ts.ts [hunk 4]
```diff
@@ -313,14 +313,23 @@ function setTaskAttributes({
...
-			if (d.image.args) {
-				task.dockerOpts.buildargs ??= {};
-				_.merge(task.dockerOpts.buildargs, d.image.args);
-			}
+		}
+		// reassign task.args so that the `--buildArg` flag takes precedence
+		// over assignments in the docker-compose.yml file (service.build.args)
+		task.args = {
+			...task.args,
+			...buildOpts.buildargs,
```
Comment (pdcastro) [inline lib/utils/compose_ts.ts:323]: This fixes issue #1053. Previously, `d.image.args` overwrote the `--buildArg` command-line option(s). Now `--buildArg` takes precedence, in line with the behavior of the standard `docker-compose` command-line tool.

File: tests/test-data/projects/docker-compose/basic/docker-compose.yml [hunk 5]
```diff
@@ -12,3 +12,5 @@ services:
     build:
       context: ./service2
       dockerfile: Dockerfile-alt
+      args:
+        - 'COMPOSE_ARG=an argument defined in the docker-compose.yml file'
```
Comment (pdcastro) [inline tests/test-data/projects/docker-compose/basic/docker-compose.yml:16]: This added variable in this test `docker-compose.yml` file allows us to test that the `--buildArg` command-line argument takes precedence.

**Review by srlowe — APPROVED** [review]

**Pull request merged**